File: Demo/VLFRunnerObjC/VFLViewController.m
//
//  VFLViewController.m
//  VLFRunnerObjC
//
//  Created by Sidharth Juyal on 09/07/2023.
//

#import "VFLViewController.h"
#import "VFLContext.h"
#import "VFLUtils.h"
#import "VFLExampleView.h"

@interface VFLViewController () {
  VFLExample _example;
  VFLContext *_context;
}
@end

@implementation VFLViewController

+ (instancetype)viewControllerWithExample:(VFLExample)example
{
  return [[VFLViewController alloc] initWithExample:example];
}

- (instancetype)initWithExample:(VFLExample)example
{
  self = [super initWithNibName:nil bundle:nil];
  if (self) {
    _example = example;
    _context = [VFLContext context];
  }
  return self;
}

- (void)viewDidLoad {
  [super viewDidLoad];
  self.title = VFLExampleGetName(_example);
  NSString *classname = VFLExampleGetClassname(_example);
  VFLExampleView *contentView = (VFLExampleView *)[[NSClassFromString(classname) alloc] initWithExample:_example];
  [_context setParentView:self.view];
  [_context addSubviewsWithBindings:NSDictionaryOfVariableBindings(contentView)];
  self.view.backgroundColor = [UIColor whiteColor];
}

- (void)viewSafeAreaInsetsDidChange
{
  [super viewSafeAreaInsetsDidChange];
  [_context removeAllConstraints];
  [_context addMetrics:vfl_CreateUIEdgeInsetsMetrics(self.view.safeAreaInsets)];
  [_context addConstraintsWithFormats:
   [NSArray arrayWithObjects:
    @"V:|-(top)-[contentView]-(bottom)-|",
    @"H:|-(left)-[contentView]-(right)-|",
    nil]];
}

@end

File: Demo/VFLRunner/Sources/VFLViewController.swift
//
//  VFLViewController.swift
//  VFLRunner
//
//  Created by Sidharth Juyal on 31/12/2022.
//

import UIKit
import VFL

class VFLViewController: UIViewController {
  private lazy var vfl = VFL(view)
  private let contentView: UIView
  
  init(contentView: UIView, title: String) {
    self.contentView = contentView
    super.init(nibName: nil, bundle: nil)
    self.title = title
  }
  
  required init?(coder: NSCoder) {
    fatalError("init(coder:) has not been implemented")
  }
  
  override func viewDidLoad() {
    super.viewDidLoad()
    vfl
      .add(subview: contentView, name: "contentVw")
      .addMetrics(view.safeAreaInsets.metrics)
      .storeConstraints(
        formats: [
          "V:|-(top)-[contentVw]-(bottom)-|",
          "H:|-(left)-[contentVw]-(right)-|"
        ],
        name: "constraints"
      )
      .applyConstraints(name: "constraints")
    
    view.backgroundColor = .white
  }
  
  override func viewSafeAreaInsetsDidChange() {
    super.viewSafeAreaInsetsDidChange()
    vfl
      .removeAll()
      .addMetrics(view.safeAreaInsets.metrics)
      .applyConstraints(name: "constraints")
  }
}
